### PR TITLE
add consent timing

### PR DESCRIPTION
**Pull request by sndrs**
## What does this change?

- adds consent timing 

correlate to https://github.com/guardian/frontend/pull/23185

### CODE REVIEW

File: src/web/browser/ga/ga.ts [hunk 1]
```diff
@@ -189,3 +189,26 @@ export const trackNonClickInteraction = (actionName: string): void => {
...
+        return;
+    }
+
+    if (window.performance?.now) {
+        ga(
+            send,
+            'timing',
+            timingCategory,
+            timingVar,
+            Math.round(window.performance.now()),
```
Comment (mxdvl) [inline src/web/browser/ga/ga.ts:210]: Bold. Not even a `ceil`.

**Review by mxdvl — APPROVED** [review]
👍 LGTM

File: src/web/browser/ga/ga.ts [hunk 2]
```diff
@@ -189,3 +189,26 @@ export const trackNonClickInteraction = (actionName: string): void => {
...
+    timingVar: any,
+    timingLabel: string,
+): void => {
+    const { ga } = window;
+
+    if (!ga) {
+        return;
+    }
+
+    if (window.performance?.now) {
```
Comment (liywjl) [inline src/web/browser/ga/ga.ts:204]: is this to check that a performance test is in progress?

File: src/web/components/App.tsx [hunk 3]
```diff
@@ -330,6 +331,21 @@ export const App = ({ CAPI, NAV }: Props) => {
             };
             injectPrivacySettingsLink(); // manually updates the footer DOM because it's not hydrated
 
+            // keep this in sync with static/src/javascripts/boot.js in frontend
+            let recordedConsentTime = false;
+            cmp.willShowPrivacyMessage().then((willShow) => {
+                onConsentChange(() => {
+                    if (!recordedConsentTime) {
+                        trackPerformance(
```
Comment (liywjl) [inline src/web/components/App.tsx:339]: we already have a performance measurement lib for components `/src/web/browser/initPerf.ts`

**Review by liywjl — COMMENTED** [review]
I am unsure if using `browser/initPerf.ts` would be preferable in this case rather than having your own implementation.

File: src/web/components/App.tsx (hunk 3 above)
Comment (sndrs) [reply src/web/components/App.tsx:340]: does it send timings to google analytics?

File: src/web/browser/ga/ga.ts (hunk 2 above)
Comment (sndrs) [reply src/web/browser/ga/ga.ts:204]: it makes sure the [API](https://developer.mozilla.org/en-US/docs/Web/API/Performance) exists in the browser

File: src/web/browser/ga/ga.ts (hunk 1 above)
Comment (sndrs) [reply src/web/browser/ga/ga.ts:210]: it's what the GA docs use https://developers.google.com/analytics/devguides/collection/analyticsjs/user-timings#measuring_time

File: src/web/components/App.tsx [hunk 4]
```diff
@@ -330,6 +331,21 @@ export const App = ({ CAPI, NAV }: Props) => {
             };
             injectPrivacySettingsLink(); // manually updates the footer DOM because it's not hydrated
 
+            // keep this in sync with static/src/javascripts/boot.js in frontend
```
[the remark this answers is not in the document]
Comment (sndrs) [reply src/web/components/App.tsx:334]: how do you mean?

File: src/web/browser/ga/ga.ts (hunk 2 above)
Comment (liywjl) [reply src/web/browser/ga/ga.ts:204]: I feel like its a given these days https://caniuse.com/?search=performance but why not 🙂

File: src/web/components/App.tsx (hunk 3 above)
Comment (liywjl) [reply src/web/components/App.tsx:340]: no it doesn't 🙂  plus I noticed subtle differences, so all is good 👍

**Review by liywjl — APPROVED** [review]
lgtm

File: src/web/components/App.tsx (hunk 3 above)
Comment (sndrs) [reply src/web/components/App.tsx:340]: cheers :) 👍

File: src/web/components/App.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (mxdvl) [reply src/web/components/App.tsx:334]: Some marker that can easily be found in the other file:

8b6c7932-944f-4fee-a1c1-4563cf4feb79
https://www.uuidgenerator.net/version4

I’ve seen this being used elsewhere:

https://github.com/guardian/frontend/commit/8cb37e5f31b60f2dd4496794a4348f251c3bdc17

[the remark this answers is not in the document]
Comment (sndrs) [reply src/web/components/App.tsx:334]: ooooh! i had no idea what that meant – such a good idea!

[the remark this answers is not in the document]
Comment (mxdvl) [reply src/web/components/App.tsx]: Yes, as searching for a string in all the codebase takes less than a second and you’re 99.999…% sure it’s where you’re meant to look.

e.g.: https://github.com/search?q=org%3Aguardian+432b3a46-90c1-4573-90d3-2400b51af8d0&type=code

**Pull request merged**